src/cmd_line/subparsers/substitute.ts

```diff
@@ -128,40 +128,67 @@ function parseCount(scanner: Scanner): number {
  */
 export function parseSubstituteCommandArgs(args: string): node.SubstituteCommand {
   try {
+    let searchPattern: string | undefined;
+    let replaceString: string;
+    let flags: number;
+    let count: number;
+
+    if (!args) {
+      // special case for :s
+      return new node.SubstituteCommand({
+        pattern: undefined,
+        replace: '', // ignored in this context
```

Comment:
I'm thinking if we can use the same rule for `replace` as `pattern`, use previous state when it's `undefined`, instead of using `pattern's value to control if `replace` should be used.

Reply:
Ah.  I remember wanting to do this, but totally forgot that there was a reason why I couldn't.  The replacement actually works differently than the pattern in vim.  Contrary to pattern, "empty replacement" is always interpreted by vim as "a replacement with nothing", no matter what the previous replacement was ;) (EXCEPT for the ":s" case, which works differently).

So, for example, if you do a ":s/this/that/", and then do a ":s/" or a ":s//", ":s///g", etc...  it will always delete occurrences of "this".

I'm glad you brought this up, because it seems I missed this case in the tests :-!  I'll definitely add it :).

- [ ] add tests for empty replacement string deleting, regardless of previous state

Reply:
Hi @rebornix 

I added a new test to check for all the variations of empty strings.  Please let me know if reading the new test resolves this item.

With that in mind, I do understand that this totally isn't the easiest code to read.

Now that all the different sequences of events are pretty well covered by tests, if you have any other ideas you'd like me to try in line with this, I'll be happy to try them, and we can be sure it'll have the same behavior :).